Earthfile

```diff
@@ -39,241 +63,273 @@ setup:
+    libboost-thread-dev libboost-test-dev \
+    libsqlite3-dev libtbb-dev libz-dev libedit-dev \
+    python3 python3-pip python3-venv \
+    llvm-14 llvm-14-dev llvm-14-tools clang-14 ros-dev-tools
+
+###############################################################################
+### Sources Stage
+# This stage is responsible for fetching the ROS 2 workspace sources.
+###############################################################################
 sources:
   FROM +setup
-  COPY +spaceros-artifacts/ros2.repos ros2.repos
-  COPY +spaceros-artifacts/excluded-pkgs.txt excluded-pkgs.txt
-  COPY +spaceros-artifacts/exact.repos exact.repos
-  RUN mkdir src
-  RUN vcs import --shallow --retry 3 --input exact.repos src
-  SAVE ARTIFACT src AS LOCAL src
 
-vcs-exact:
-  COPY +spaceros-artifacts/${SPACEROS_DIR}/exact.repos exact.repos
-  SAVE ARTIFACT exact.repos AS LOCAL exact.repos
+  RUN apt install -y python3-vcstool \
+    && mkdir src -p \
+    && vcs import --retry 3 src < output.repos \
+    && vcs export --exact src > exact.repos
+
+  # Save artifacts to be  used
+  SAVE ARTIFACT exact.repos
+  SAVE ARTIFACT src
 
+###############################################################################
+### Rosdep Stage
+# This stage is responsible for installing the system dependencies required
+# for the ROS 2 workspace.
+###############################################################################
 rosdep:
-  FROM +sources
-  COPY src src/
+  FROM +pre-installation
+
+  # Rosdep updates
+  RUN apt-get install -y python3-rosdep \
+    && rosdep init \
+    && rosdep update
+
+  # Copy Repos file
+  COPY +sources/src ./src
+  COPY excluded-pkgs.txt excluded-deps.txt ./
 
   # Install system package dependencies using rosdep
-  RUN sudo rosdep init && rosdep update
-  RUN --mount=type=cache,mode=0777,target=/var/cache/apt,sharing=locked,id=cache_apt_cache \
-      --mount=type=cache,mode=0777,target=/var/lib/apt,sharing=locked,id=lib_apt_cache \
-      sudo apt-get update && \
-      rosdep update && \
-      rosdep install -y \
+  RUN rosdep install -y \
         --from-paths src --ignore-src \
+        --simulate \
         --rosdistro ${ROS_DISTRO} \
         # `urdfdom_headers` is cloned from source, however rosdep can't find it.
         # It is because package.xml manifest is missing. See: https://github.com/ros/urdfdom_headers
         # Additionally, IKOS must be excluded as per: https://github.com/space-ros/docker/issues/99
-        --skip-keys "$(tr '\n' ' ' < 'excluded-pkgs.txt') urdfdom_headers ikos"
-  RUN rm excluded-pkgs.txt
+        --skip-keys "$(tr '\n' ' ' < 'excluded-pkgs.txt') urdfdom_headers ikos" > rosdeps.txt
+
+  # Process rosdeps.txt to a shell script
+  RUN touch rosdeps.sh \
+        && echo "#!/bin/bash" > rosdeps.sh \
+        && echo "apt-get update" >> rosdeps.sh \
+        && echo "apt-get install -y \\" >> rosdeps.sh \
+        && grep -v -F -f excluded-deps.txt rosdeps.txt | sed 's/^/  /' >> rosdeps.sh \
```

Comment:
Might be missing something but can you elaborate on where the excluded-deps.txt is coming from? Just manually generated based on things we don't need?

Reply:
Ah, I think some of these are definitely needed in `build-testing`, since

This is what I get testing locally:
```
- rosidl_typesupport_interface clang_tidy.xunit.missing_result
  <<< error message
    The test did not generate a result file:
    
    Could not find 'clang-tidy' / 'clang-tidy-10' / 'clang-tidy-11' / 'clang-tidy-6.0' executable
  >>>
build/rosidl_typesupport_interface/test_results/rosidl_typesupport_interface/cobra-autosar.xunit.xml: 1 test, 1 error, 0 failures, 0 skipped
- rosidl_typesupport_interface cobra-autosar.xunit.missing_result
  <<< error message
    The test did not generate a result file.
  >>>
build/rosidl_typesupport_interface/test_results/rosidl_typesupport_interface/cppcheck.xunit.xml: 1 test, 1 error, 0 failures, 0 skipped
- rosidl_typesupport_interface cppcheck.xunit.missing_result
  <<< error message
    The test did not generate a result file:
    
    Could not find 'cppcheck' executable
  >>>

Summary: 325 tests, 30 errors, 13 failures, 3 skipped
```

Reply:
This is the [`excluded-deps.txt`](https://github.com/space-ros/space-ros/pull/186/files/23538d13f44a22becdf5acc6275f20d9dd604041#diff-a0bf546ac9698b4feb06fd4aaddc5c84823f269740e2e780df79ca6c5c13c5c3). If should be in the root of the repository

Reply:
I didn't give much time to build-test and I should be taking a look at the original build to understand.

Reply:
Yes, build-test is... definitely the slowest part, but it's going now!